src/java/org/apache/cassandra/cql3/QueryProcessor.java

```diff
@@ -230,7 +235,16 @@ private QueryProcessor()
     @VisibleForTesting
     public void evictPrepared(MD5Digest id)
     {
-        preparedStatements.invalidate(id);
+        // Caffeine key collision of concurrent invalidation and computation has to be avoided DSP-24330
+        try
+        {
+            while (preparedStatementsUpdateLock.putIfAbsent(id, true) != null);
```

Comment:
It looks like a concurrent hash map has been used to create something like a striped lock (even though it's not actually a lock).
The thread that puts the statement into the cache (via the `get` operation in https://github.com/datastax/cassandra/pull/1257/files#diff-3ce9f6c4626262d6a4be26ecf3a57ef11fdb55537448def0c736115417487115R769) won't be able to execute until the `id` is removed at https://github.com/datastax/cassandra/pull/1257/files#diff-3ce9f6c4626262d6a4be26ecf3a57ef11fdb55537448def0c736115417487115R246.

Reply:
Correct. And the busy spin is intentional to avoid the Thread un/parking overhead.